vessel-runtime/src/test/java/com/textnow/android/vessel/VesselCacheUsageTest.kt

```diff
@@ -0,0 +1,295 @@
+        basicVessel.delete(Data1::class)
+        val read2 = basicVessel.get(Data1::class)
+
+        val profData = vessel.profileData
+
+        assertThat(cache!!.get<Data1>(Data1::class.qualifiedName!!)).isEqualTo(VesselImpl.nullValue)
+        assertThat(data1).isEqualTo(read1)
+        assertThat(null).isEqualTo(read2)
+        assertThat(profData?.hitCountOf(Span.DELETE_FROM_DB)).isEqualTo(1)
+        assertThat(profData?.hitCountOf(Event.CACHE_HIT_READ)).isEqualTo(2)
+    }
+}
+
+@Config(sdk = [Build.VERSION_CODES.P])
+@RunWith(AndroidJUnit4::class)
+class AsyncVesselCacheUsageTest: BaseVesselCacheUsageTest(true)
+
+@Config(sdk = [Build.VERSION_CODES.P])
+@RunWith(AndroidJUnit4::class)
+class SyncVesselCacheUsageTest: BaseVesselCacheUsageTest(false)
+
+@Config(sdk = [Build.VERSION_CODES.P])
+@RunWith(AndroidJUnit4::class)
+class AsyncOnlyVesselCacheUsageTest: BaseVesselTest<TestCache>(TestCache(), true) {
+    @Test
+    fun `replace with new type is cached`() = runBlocking {
+        vessel.set(data1)
+        // Note - this will deadlock if using @Rule with InstantTaskExecutorRule.  All other
+        // suspend functions still work fine
+        // Perhaps some variation on https://issuetracker.google.com/issues/120854786
+        vessel.replace(Data1::class, data2)
+
+        val profData = vessel.profileData
+
+        assertThat(profData?.hitCountOf(Span.REPLACE_IN_DB)).isEqualTo(1)
+        assertThat(cache!!.get<Data1>(Data1::class.qualifiedName!!)).isEqualTo(VesselImpl.nullValue)
+        assertThat(cache.get<Data2>(Data2::class.qualifiedName!!)).isEqualTo(data2)
+    }
+
+    @Test
+    fun `replace with same type is equivalent to set`() = runBlocking {
+        vessel.set(data1)
+        // Note:  This will deadlock if using InstantTaskExecutorRule
+        vessel.replace(Data1::class, Data1(2))
+
+        val profData = vessel.profileData
+
+        assertThat(profData?.hitCountOf(Span.WRITE_TO_DB)).isEqualTo(2)
+        assertThat(cache!!.get<Data1>(Data1::class.qualifiedName!!)).isEqualTo(Data1(2))
+    }
+
+    @Test
+    fun `already cached data is not replaced in database`() = runBlocking {
+        vessel.set(data1)
+        // Note:  This will deadlock if using InstantTaskExecutorRule
+        vessel.replace(Data1::class, Data2(2))
+        vessel.replace(Data1::class, Data2(2))
+
+        val profData = vessel.profileData
+
+        assertThat(profData?.hitCountOf(Span.WRITE_TO_DB)).isEqualTo(1)
+        assertThat(profData?.hitCountOf(Event.CACHE_HIT_REPLACE)).isEqualTo(1)
+        assertThat(cache!!.get<Data1>(Data1::class.qualifiedName!!)).isEqualTo(VesselImpl.nullValue)
+    }
+
+}
+
+
+@Config(sdk = [Build.VERSION_CODES.P])
+@RunWith(AndroidJUnit4::class)
+class ObserversVesselCacheUsageTest: BaseVesselTest<TestCache>(TestCache()) {
+    // Note:  This will deadlock certain transactional DAO methods, such as replace (but not set/delete)
+    @get:Rule()
+    val executor = InstantTaskExecutorRule()
```

Comment:
You should be setting up the dispatchers. See FlowTest.

Reply:
I wonder if this could be considered an instrumented test as it runs under Robolectric.
Same as your LiveDataTest.  If this test code can be kept simpler that would be ideal - I don't think the custom dispatcher is required.

The docs you linked indicate a dispatcher is only set up for local tests:
>Note: This only applies to local unit tests. You should not replace the Main dispatcher in [instrumented tests](https://developer.android.com/training/testing/instrumented-tests) where the real UI thread is available.